### PR TITLE
Pre process final

### PR DESCRIPTION
**Pull request by carlhiggs**
This branch includes a preprocess folder of code for calculation of preliminary resources given input data sets (which are not included --- I plan to load a 'data' folder up on CloudStor containing these GHS and OSM resources which can be downloaded to merge with the repo process/data folder.  

The caveat is, this is used for for calculation of basic resources as previously delivered --ie. OSM destinations (and POS entry points, for any and large).  However, it doesn't yet include additional code to incorporate custom destinations (Maiduguri) or regular PT data; and I haven't yet tested out that all that will be required for evaluating access to POS and regular PT data has been done.  For example, the OSM type destinations were pre-associated with their nearest node locations -- I can't recall if that information is made use of in how we ended up using the code; if the analysis assumes these fields are present for destinations, then we'll have to either add them for POS and PT, or combine those destinations altogether with the OSM ones.

At a basic level though, the resources are prepared enough to make a start on analyses I believe, and I think the analysis process code should allow us to post hoc go back and analyse the other destinations as we add them.

tl/dr: I expect updates will be made to the pre-processing code, as required!

### CODE REVIEW

**Comment by gboeing** [comment]
@carlhiggs I see there's also a [pre_process](https://github.com/gboeing/global-indicators/tree/pre_process) branch... shall we delete that one? Is it outdated?

**Comment by carlhiggs** [comment]
Hi @gboeing , yup we shall delete it --- when i first created it (and related to the other issue i posted earlier) i overlooked that I wasn't on the master branch.  I attempted to merge the master into it post hoc, but it didn't seem to resolve the conflicts so it made more sense to just create a new branch and do things properly.  I'm in the process of cleaning up folders on my computer (as the pre-processing work I did the past few days was actually on that branch), and once I've sorted things out so I know i won't lose data if i switch branches I'll delete the other one.

**Comment by gboeing** [comment]
> At a basic level though, the resources are prepared enough to make a start on analyses I believe, and I think the analysis process code should allow us to post hoc go back and analyse the other destinations as we add them.
> 
> tl/dr: I expect updates will be made to the pre-processing code, as required!

Ok, I'll hold off on merging this with master until I get the official word from you.

**Comment by carlhiggs** [comment]
As per discussion on e-mail, I'm working on updating the docker image definition and making sure the validation report documentation works as intended when run back in January (so merge is still pending while I run through the code)  @gboeing

**Comment by carlhiggs** [comment]
Hi @gboeing  @shiqin-liu @nicholas-david @duelran , 

I think this branch is good to be merged in now!  (edits -- i'm very tired, messed up git lingo )

I've updated the pre-process code so it now is able to generate the preliminary data validation report --- I'll attach Maiduguri's as a proof of concept here.  It requires all cities defined in the configuration file to have been processed to successfully run, and if attempt is made to run it without doing this the user will be prompted (because the key all cities summary table won't exist).  The summary tables used to be manually constructed from CSV, but now are built using some customised latex generated from Pandas --- I'll attach here in case of use for our other paper(s) (e.g. it contains index rows for continent, country and city --- so the data columns could be repurposed for other uses).  The python code for that is in the _all_cities_summary_tex_tables_for_report.py script.

[global_liveability_maiduguri.pdf](https://github.com/gboeing/global-indicators/files/5304146/global_liveability_maiduguri.pdf)

zipped tex file (as it wouldn't allow upload of the tex file itself -- I didn't include this tex file in the repository as its subject to change pending configuration and data)
[cities_data.zip](https://github.com/gboeing/global-indicators/files/5304148/cities_data.zip)

cheers,

Carl

File: docker/Dockerfile [hunk 1]
```diff
@@ -37,6 +37,33 @@ RUN conda config --set show_channel_urls true && \
     conda info --all && \
     conda list
 
+# install tools for using OpenStreetMap
+RUN apt-get update && apt-get install -y osm2pgsql osmctools
```
Comment (gboeing) [inline docker/Dockerfile]: Need to put all the RUN commands before line 68 into a single RUN command (see lines 31-38 as example) to prevent image bloat.

File: process/collaborator_report/_render.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+make clean
+make html
+cp -rT _build/html ../maps/
+make latexpdf
```
Comment (gboeing) [inline process/collaborator_report/_render.sh]: Add blank line at end of file.

File: process/pre_process/timeTask.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  # "application" code
+  logger.debug("debug message")
+  logger.info("info message")
+  logger.warn("warn message")
+  logger.error("error message")
+  logger.critical("critical message")
+
+# Write the output to a text file
+text_file.write("OID = %s; count = %s" % (max(data, key = data.get), data[max(data, key = data.get)]))
+text_file.close()
```
Comment (gboeing) [inline process/pre_process/timeTask.py]: Add a blank line at end of file.

**Review by gboeing — CHANGES_REQUESTED** [review]
@carlhiggs looking great. A couple minor requests included in this review...

**Comment by carlhiggs** [comment]
Thanks @gboeing I just amended the docker image file as suggested (main RUN all in one --- the run calls to python and ipython are still separate, but I don't expect that matters as was that way in original, I think).  I gave it a test run and re-built the pdf report successfully.  I haven't re-run all the other code since previous runs, but can see no reason why this shouldn't also work.  Perhaps before final release we can do a full run through all code using the master branch.  ----- So I believe this pull request should be good to merge if all agree!

File: process/pre_process/timeTask.py (hunk 3 above)
Comment (carlhiggs) [reply process/pre_process/timeTask.py]: I removed this file; it was unused in the code in the end.

File: process/collaborator_report/_render.sh (hunk 2 above)
Comment (carlhiggs) [reply process/collaborator_report/_render.sh]: Done!

File: docker/Dockerfile (hunk 1 above)
Comment (carlhiggs) [reply docker/Dockerfile]: This has also been done -- with exception of the calls to python and ipython (as per main comment below; I think these were in original docker file, so assume intended to be separate).

**Comment by gboeing** [comment]
> the run calls to python and ipython are still separate, but I don't expect that matters as was that way in original, I think

Correct!

> Perhaps before final release we can do a full run through all code using the master branch.

Yep, I agree!

**Review by gboeing — APPROVED** [review]

File: docker/Dockerfile [hunk 4]
```diff
@@ -35,7 +35,31 @@ RUN conda config --set show_channel_urls true && \
...
+    polyglossia fontspec cmap gnu-freefont upquote oberdiek latexmk \
+    caption makecell multirow changepage \
+    --repository=http://mirror.ctan.org/systems/texlive/tlnet \
+    && /root/bin/fmtutil-sys --all && \
+    # Install additional LaTeX packages
+    # installing seperately to avoid error associated with multirow && \
+    # install the make build tools, for compiling sphinx documentation
+    apt-get update && apt-get install -y make   && \
+    # install the contextily package using pip, required for validation report basemaps
+    RUN pip install contextily
```
Comment (gboeing) [inline docker/Dockerfile]: One last thing... `contextily` is available via conda-forge: https://anaconda.org/conda-forge/contextily Can we just add this to `requirements.txt` instead of having a separate pip install just for it?

If not, can we remove the `RUN` part here so `pip install` operates within the single `RUN` command above?

**Review by gboeing — CHANGES_REQUESTED** [review]

**Review by gboeing — APPROVED** [review]

File: docker/Dockerfile (hunk 4 above)
Comment (carlhiggs) [reply docker/Dockerfile]: Thanks Geoff -- contextily is written in the requirements.txt, but it seems it is nevertheless required to be installed using pip; when I didn't do this, it asked for the package to be installed (and once installed using pip, it worked).  I accidentally missed removing that RUN command when concatenating the build into a single run --- that's now fixed and pushed.  Cheers,  Carl
